packages/wrangler/src/__tests__/type-generation.test.ts

```diff
@@ -1,11 +1,76 @@
 import * as fs from "fs";
 import * as TOML from "@iarna/toml";
+import {
+	constructType,
+	constructTypeKey,
+	isValidIdentifier,
+} from "../type-generation";
 import { dedent } from "../utils/dedent";
 import { mockConsoleMethods } from "./helpers/mock-console";
 import { runInTempDir } from "./helpers/run-in-tmp";
 import { runWrangler } from "./helpers/run-wrangler";
 import type { EnvironmentNonInheritable } from "../config/environment";
 
+describe("isValidIdentifier", () => {
+	it("should return true for valid identifiers", () => {
+		expect(isValidIdentifier("valid")).toBe(true);
+		expect(isValidIdentifier("valid123")).toBe(true);
+		expect(isValidIdentifier("valid_123")).toBe(true);
+		expect(isValidIdentifier("valid_123_")).toBe(true);
+		expect(isValidIdentifier("_valid_123_")).toBe(true);
+		expect(isValidIdentifier("_valid_123_")).toBe(true);
+		expect(isValidIdentifier("$valid")).toBe(true);
+		expect(isValidIdentifier("$valid$")).toBe(true);
+	});
+
+	it("should return false for invalid identifiers", () => {
+		expect(isValidIdentifier("123invalid")).toBe(false);
+		expect(isValidIdentifier("invalid-123")).toBe(false);
+		expect(isValidIdentifier("invalid 123")).toBe(false);
+	});
+});
+
+describe("constructTypeKey", () => {
+	it("should return a valid type key", () => {
+		expect(constructTypeKey("valid")).toBe("valid");
+		expect(constructTypeKey("valid123")).toBe("valid123");
+		expect(constructTypeKey("valid_123")).toBe("valid_123");
+		expect(constructTypeKey("valid_123_")).toBe("valid_123_");
+		expect(constructTypeKey("_valid_123_")).toBe("_valid_123_");
+		expect(constructTypeKey("_valid_123_")).toBe("_valid_123_");
+		expect(constructTypeKey("$valid")).toBe("$valid");
+		expect(constructTypeKey("$valid$")).toBe("$valid$");
+
+		expect(constructTypeKey("123invalid")).toBe('"123invalid"');
+		expect(constructTypeKey("invalid-123")).toBe('"invalid-123"');
+		expect(constructTypeKey("invalid 123")).toBe('"invalid 123"');
+	});
+});
+
+describe("constructType", () => {
+	it("should return a valid type", () => {
+		expect(constructType("valid", "string")).toBe("valid: string;");
+		expect(constructType("valid123", "string")).toBe("valid123: string;");
+		expect(constructType("valid_123", "string")).toBe("valid_123: string;");
+		expect(constructType("valid_123_", "string")).toBe("valid_123_: string;");
+		expect(constructType("_valid_123_", "string")).toBe("_valid_123_: string;");
+		expect(constructType("_valid_123_", "string")).toBe("_valid_123_: string;");
+
+		expect(constructType("123invalid", "string")).toBe('"123invalid": string;');
+		expect(constructType("invalid-123", "string")).toBe(
+			'"invalid-123": string;'
+		);
+		expect(constructType("invalid 123", "string")).toBe(
+			'"invalid 123": string;'
+		);
+
+		expect(constructType("valid", 'a"', false)).toBe('valid: "a\\"";');
+		expect(constructType("valid", "a\\", false)).toBe('valid: "a\\"";');
```

Comment:
I think there an extra `"` here
```suggestion
		expect(constructType("valid", "a\\", false)).toBe('valid: "a\\";');
```

Reply:
Good catch!, thanks Not sure how I missed this locally 🤔